src/tap/src/tapcell.cpp

```diff
@@ -33,41 +33,1306 @@
+       = macro_outlines.begin();
+       iter != macro_outlines.end();
+       ++iter) {
+    std::pair<int, int> k = iter->first;
+    vector<int> all_rows = macro_outlines[k];
+    vector<int> new_rows;
+    new_rows.push_back(all_rows.front());
+    for (int i = 1; i < all_rows.size() - 1; i++) {
+      if (all_rows[i] + 1 == all_rows[i + 1]) {
+        continue;
+      }
+      new_rows.push_back(all_rows[i]);
+      new_rows.push_back(all_rows[i + 1]);
+    }
+    new_rows.push_back(all_rows.back());
+    if (macro_outlines_array.find(k) == macro_outlines_array.end()) {
+      macro_outlines_array.insert(
+          std::pair<std::pair<int, int>, vector<int>>(k, new_rows));
+    } else {
+      macro_outlines_array[k].insert(
+          macro_outlines_array[k].end(), new_rows.begin(), new_rows.end());
+    }
+  }
+
+  return macro_outlines_array;
+}
+
+// // proc to detect if blockage overlaps with row
+bool Tapcell::overlaps(odb::dbBox* blockage,
+                       odb::dbRow* row,
+                       int& halo_x,
+                       int& halo_y)
+{
+  odb::Rect rowBB;
+  row->getBBox(rowBB);
+
+  // check if Y has overlap first since rows are long and skinny
+  int blockage_lly = blockage->yMin() - halo_y;
+  int blockage_ury = blockage->yMax() + halo_y;
+  int row_lly = rowBB.yMin();
+  int row_ury = rowBB.yMax();
+
+  if (blockage_lly >= row_ury || row_lly >= blockage_ury) {
+    return false;
+  }
+
+  int blockage_llx = blockage->xMin() - halo_x;
+  int blockage_urx = blockage->xMax() + halo_x;
+  int row_llx = rowBB.xMin();
+  int row_urx = rowBB.xMax();
+
+  if (blockage_llx >= row_urx || row_llx >= blockage_urx) {
+    return false;
+  }
+
+  return true;
+}
+
+vector<odb::dbBox*> Tapcell::findBlockages()
+{
+  odb::dbBlock* block = db_->getChip()->getBlock();
+  vector<odb::dbBox*> blockages;
+
+  for (auto&& inst : db_->getChip()->getBlock()->getInsts()) {
+    if (inst->isBlock()) {
+      if (!inst->isPlaced()) {
+        logger_->warn(utl::TAP, 32, "Macro {} is not placed.", inst->getName());
+        continue;
+      }
+      blockages.push_back(inst->getBBox());
+    }
+  }
+
+  return blockages;
+}
+
+int Tapcell::makeSiteLoc(int x, double site_x, int dirc, int& offset)
+{
+  auto site_int = double(x - offset) / site_x;
+  if (dirc == 1) {
+    site_int = ceil(site_int);
+  } else {
+    site_int = floor(site_int);
+  }
+  site_int = int(site_int * site_x + offset);
+  return site_int;
+}
+
+void Tapcell::buildCell(odb::dbBlock* block,
+                        odb::dbMaster* master,
+                        odb::dbOrientType orientation,
+                        int x,
+                        int y,
+                        string prefix)
+{
+  if (x < 0 || y < 0) {
+    return;
+  }
+  string name = prefix + std::to_string(phy_idx_);
+  odb::dbInst* inst = odb::dbInst::create(block, master, name.c_str());
+  if (inst == nullptr) {
+    return;
+  }
```

Comment:
Is it really ok to return or is this an error of some sort?  It can only really happen with a duplicate name.